purchases/src/test/java/com/revenuecat/purchases/attributes/SubscriberAttributeTests.kt

```diff
@@ -0,0 +1,123 @@
+package com.revenuecat.purchases.attributes
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.revenuecat.purchases.timeInSeconds
+import com.revenuecat.purchases.utils.DateProvider
+import org.assertj.core.api.Assertions.assertThat
+import org.json.JSONException
+import org.json.JSONObject
+import org.junit.Assert.fail
+import org.junit.Test
+import org.junit.runner.RunWith
+import java.util.Date
+
+@RunWith(AndroidJUnit4::class)
+class SubscriberAttributeTests {
+
+    @Test
+    fun `timestamp is correctly set`() {
+        val now = Date()
+        val dateProvider = object : DateProvider {
+            override val now: Date
+                get() = now
+        }
+        val subscriberAttribute =
+            SubscriberAttribute("email", "un@email.com", dateProvider = dateProvider)
+        assertThat(subscriberAttribute.setTime).isEqualTo(now)
+    }
+
+    @Test
+    fun `toJSONObject works`() {
+        val now = Date()
+        val emailKey = "email"
+        val emailValue = "un@email.com"
+        val toJSONObject = SubscriberAttribute(emailKey, emailValue, setTime = now).toJSONObject()
+        assertThat(toJSONObject.getString(JSON_NAME_KEY)).isEqualTo(emailKey)
+        assertThat(toJSONObject.getString(JSON_NAME_VALUE)).isEqualTo(emailValue)
+        assertThat(toJSONObject.getLong(JSON_NAME_SET_TIME)).isEqualTo(now.time)
+    }
+
+    @Test
+    fun `JSONObject constructor works`() {
+        val now = Date()
+        val emailKey = "email"
+        val emailValue = "un@email.com"
+        val jsonObject = JSONObject().apply {
+            put(JSON_NAME_KEY, emailKey)
+            put(JSON_NAME_VALUE, emailValue)
+            put(JSON_NAME_SET_TIME, now.time)
+            put(JSON_NAME_IS_SYNCED, false)
+        }
+        val subscriberAttribute = SubscriberAttribute(jsonObject)
+        assertThat(subscriberAttribute.key.serverValue).isEqualTo(emailKey)
+        assertThat(subscriberAttribute.value).isEqualTo(emailValue)
+        assertThat(subscriberAttribute.setTime).isEqualTo(now)
+    }
+
+    @Test
+    fun `JSONObject missing values`() {
+        val emailKey = "email"
+        val emailValue = "un@email.com"
+        val jsonObject = JSONObject().apply {
+            put(JSON_NAME_KEY, emailKey)
+            put(JSON_NAME_VALUE, emailValue)
+        }
+        try {
+            SubscriberAttribute(jsonObject)
+            fail("Should give exception")
+        } catch (e: JSONException) {
+            assert(true)
+        } catch (e: Exception) {
+            fail("Should give JSONException")
+        }
```

Comment:
Could this block be made more expressive with [something like this](https://kotlinlang.org/api/latest/kotlin.test/kotlin.test/assert-fails-with.html)?

Reply:
I love it!